File: algorithm/LeetCode/Easy/0933-number-of-recent-calls/0933-number-of-recent-calls.py
from collections import deque

class RecentCounter:

    def __init__(self):
        self.counter = deque()

    def ping(self, t: int) -> int:
        period = t - 3000
        while len(self.counter) > 0 and self.counter[0] < period:
            self.counter.popleft()
        
        self.counter.append(t)
        return len(self.counter)


# Your RecentCounter object will be instantiated and called as such:
# obj = RecentCounter()
# param_1 = obj.ping(t)

File: algorithm/LeetCode/Easy/0933-number-of-recent-calls/0933-number-of-recent-calls.go
type RecentCounter struct {
    queue []int
}


func Constructor() RecentCounter {
    return RecentCounter{
        queue: []int{},
    }
}


func (this *RecentCounter) Ping(t int) int {
    targetIdx := 0
    for i := 0; i < len(this.queue); i++ {
        if len(this.queue) == 0 {
            break
        }

        if this.queue[i] >= t - 3000 {
            targetIdx = i
            break
        }
    }

    this.queue = this.queue[targetIdx:]    
    this.queue = append(this.queue, t)

    return len(this.queue)
}